packages/loader/container-loader/src/test/connectionStateHandler.spec.ts

```diff
@@ -212,19 +298,44 @@ describe("ConnectionStateHandler Tests", () => {
             "Client should be in Connected state");
     });
 
-    it("Should move to connected state on normal flow for write client, even if quorum isn't initialized at first", async () => {
-        // swap out quorumClients fn for one that returns undefined at first
+    it("Connect write first, init protocol later #1", async () => {
+        // ConnectionStateManager without initialized protocol
+        connectionStateHandler = createConnectionStateHandlerCore(
+            false, // connectedRaisedWhenCaughtUp,
+            false, // readClientsWaitForJoinSignal
+            handlerInputs,
+            deltaManagerForCatchingUp as any,
+            undefined);
+
+        connectionDetails.mode = "write";
+        assert.strictEqual(connectionStateHandler.connectionState, ConnectionState.Disconnected,
+            "Client should be in disconnected state");
+        connectionStateHandler.receivedConnectEvent(connectionDetails);
+        assert.strictEqual(connectionStateHandler.connectionState, ConnectionState.CatchingUp,
+            "Client should be in connecting state");
+
+        // init protocol
+        connectionStateHandler.initProtocol(protocolHandler);
+
+        connectionStateHandler_receivedAddMemberEvent(pendingClientId);
+
+        assert.strictEqual(connectionStateHandler.connectionState, ConnectionState.Connected,
+            "Client should be in connected state");
+    });
+
+    it("Connect write first, init protocol later #2", async () => {
         // ConnectionStateManager without initialized protocol
         connectionStateHandler = createConnectionStateHandlerCore(
             false, // connectedRaisedWhenCaughtUp,
+            false, // readClientsWaitForJoinSignal
             handlerInputs,
             deltaManagerForCatchingUp as any,
             undefined);
 
-        client.mode = "write";
+        connectionDetails.mode = "write";
         assert.strictEqual(connectionStateHandler.connectionState, ConnectionState.Disconnected,
             "Client should be in disconnected state");
-        connectionStateHandler.receivedConnectEvent(client.mode, connectionDetails);
+        connectionStateHandler.receivedConnectEvent(connectionDetails);
         assert.strictEqual(connectionStateHandler.connectionState, ConnectionState.CatchingUp,
             "Client should be in connecting state");
 
```

Comment:
Next few lines: In practice how could receivedAddMemberEvent be called before initProtocol?

Reply:
It took me a while to remember, but I think here is what I'm trying to do here:
First of all, this does not result in a call to ConnectionStateHandler, it only sets up quorum state.
There is a tiny tiny race possible, where these events happen in this order:
1.a connection is established (no "cached" mode is used, so it happens in parallel / faster than other steps)
2. some other client produces a summary
3. we get "lucky" and load from that summary as our initial snapshot
4. ConnectionStateHandler.initProtocol is called, "self" is already in the quorum.

We can prevent that case by reordering # 1 & # 3 in Container.load().

Thoughts? Meanwhile, I'll add comments for that sequence here. and in the code